server/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtHelper.java

```diff
@@ -55,13 +59,42 @@ public static Jwt decode(String token) {
         return new JwtImpl(header, claims, crypto);
     }
 
-    public static Jwt encode(CharSequence content, KeyInfo keyInfo) {
-        JwtHeader header = JwtHeaderHelper.create(keyInfo.algorithm(), keyInfo.keyId(), keyInfo.keyURL());
+    public static Jwt encode(CharSequence content, KeyInfo keyInfo, boolean addX5t) {
+        JwtHeader header;
+
+        if (addX5t && keyInfo.verifierCertificate() != null) {
+            HeaderParameters headerParameters = new HeaderParameters(keyInfo.algorithm(), keyInfo.keyId(), null);
+            headerParameters.setX5t(getX509CertThumbprint(getX509CertEncoded(X509CertUtils.parse(keyInfo.verifierCertificate())), "SHA-1"));
+            header = JwtHeaderHelper.create(headerParameters);
+        } else {
+            header = JwtHeaderHelper.create(keyInfo.algorithm(), keyInfo.keyId(), keyInfo.keyURL());
+        }
         byte[] claims = utf8Encode(content);
         byte[] crypto = keyInfo.getSigner()
-          .sign(concat(b64UrlEncode(header.bytes()), PERIOD, b64UrlEncode(claims)));
+            .sign(concat(b64UrlEncode(header.bytes()), PERIOD, b64UrlEncode(claims)));
         return new JwtImpl(header, claims, crypto);
     }
+
+    public static Jwt encode(CharSequence content, KeyInfo keyInfo) {
+        return encode(content, keyInfo, false);
+    }
+
+    public static byte[] getX509CertEncoded(X509Certificate x509Certificate) {
+        try {
+            return x509Certificate.getEncoded();
+        } catch (Exception e) {
```

Comment:
I am upset that the static analysis doesn't catch `catch (Exception ...`. We should catch `RuntimeException`, to prevent the code from potentially catching checked exceptions in the future. Never `Exception`.

Reply:
done